### ISSUE TITLE
module-level init() function?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#5792]
Now that we have a way for users to clean up after themselves using a module-level deinit() function (#5744), these questions come up:

* Should we also have a module-level init() function?
* Should it replace the implicit module-level initialization that we currently have?
* If not, what order should it be executed relative to the implicit module initialization?

**Comment by bradcray** [comment]
My answers are:

> Should we also have a module-level init() function?

I don't see any reason not to, and the symmetry seems attractive.

> Should it replace the implicit module-level initialization that we currently have?

I don't think it should because (a) module-level variable initializers will need to execute in the implicit initializer anyway and (b) if we dropped support for module-level statements, it would break our ability to sketch out code quickly, as in our single-line hello world program.

> If not, what order should it be executed relative to the implicit module initialization?

I would think we should run the implicit (current) module level initialization code first and then run a user's module-level init(), if present.

**Comment by npadmana** [comment]
@bradcray --- my $0.002 here --

I agree about the symmetry with ``deinit`` but I'm not sure I'm seeing a use case that the current module level statements don't cover. Do you have such a case in mind?  

In some sense, I think the symmetry was already broken in that there was a way to have code run at module initialization, but no way to run code at module de-initialization. If eg we had the equivalent of a ``defer`` statement, then I think (from a user perspective), you would not need a module ``deinit`` proc.

I would **strongly** support keeping module level statements, even if they turn into syntactic sugar and get (behind my back) converted in into an ``init`` function. It is just very nice when one is hacking out some piece of code (which is certainly one use case).

**Comment by bradcray** [comment]
> but I'm not sure I'm seeing a use case that the current module level statements don't cover. Do you have such a case in mind?

I don't think it adds any new power; but for programmers who like things "just so", it may be more appealing / structured to feel like all of their module-level initialization code was contained in one well-defined function (and perhaps one might envision a compiler warning that could be opted into by such users to keep their code tidy (?)).

I'm not such a programmer, though, so I'm speculating a bit here.

> I would strongly support keeping module level statements, even if they turn into syntactic sugar and get (behind my back) converted in into an init function. It is just very nice when one is hacking out some piece of code (which is certainly one use case).

I agree -- thanks for weighing in!
